File: apps/Pose-Motion/iOS/PoseMotion/Views/BenchmarkHUD.swift
import SwiftUI

/// The demo's selling point: live per-model latency, sustained FPS, and peak memory.
struct BenchmarkHUD: View {
    let stats: BenchmarkSnapshot
    let liftAvailable: Bool
    let mode: ClipFrameSource.Mode

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack(spacing: 8) {
                Image(systemName: "cpu")
                    .font(.system(size: 11, weight: .bold))
                Text("On-device · Melange")
                    .font(.system(size: 11, weight: .bold, design: .rounded))
                Spacer()
                Text(mode == .benchmark ? "sustained" : "realtime")
                    .font(.system(size: 10, weight: .semibold, design: .rounded))
                    .foregroundStyle(Theme.textSecondary)
            }
            .foregroundStyle(Theme.accent)

            VStack(spacing: 4) {
                HStack {
                    Text("memory")
                        .font(.system(size: 11, weight: .medium, design: .rounded))
                        .foregroundStyle(Theme.textSecondary)
                    Spacer()
                    Text(String(format: "%.0f MB · peak %.0f", stats.memoryMB, stats.peakMemoryMB))
                        .font(.system(size: 11, weight: .medium, design: .rounded))
                        .monospacedDigit()
                        .foregroundStyle(Theme.textPrimary)
                }
                latencyRow("YOLO26n", stats.detectMs)
                latencyRow("RTMPose-s", stats.poseMs)
                if liftAvailable {
                    latencyRow("3D lift", stats.liftMs)
                }
                latencyRow("pipeline", stats.totalMs, emphasized: true)
            }
        }
        .padding(12)
        .frame(width: 220)
        .background(.ultraThinMaterial, in: RoundedRectangle(cornerRadius: 14, style: .continuous))
    }

    private func latencyRow(_ label: String, _ ms: Double, emphasized: Bool = false) -> some View {
        HStack {
            Text(label)
                .font(.system(size: 11, weight: emphasized ? .bold : .medium, design: .rounded))
                .foregroundStyle(emphasized ? Theme.textPrimary : Theme.textSecondary)
            Spacer()
            Text(ms > 0 ? String(format: "%.1f ms", ms) : "—")
                .font(.system(size: 11, weight: emphasized ? .bold : .medium, design: .rounded))
                .monospacedDigit()
                .foregroundStyle(emphasized ? Theme.accent : Theme.textPrimary)
        }
    }
}

File: apps/Pose-Motion/Android/app/src/main/java/ai/zetic/demo/posemotion/ui/BenchmarkHud.kt
package ai.zetic.demo.posemotion.ui

import ai.zetic.demo.posemotion.pipeline.BenchmarkSnapshot
import ai.zetic.demo.posemotion.video.ClipFrameSource
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Memory
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import java.util.Locale

/** The demo's selling point: live per-model latency, sustained FPS, and peak memory. */
@Composable
fun BenchmarkHud(
    stats: BenchmarkSnapshot,
    liftAvailable: Boolean,
    mode: ClipFrameSource.Mode,
) {
    Column(
        modifier = Modifier
            .width(220.dp)
            .background(Theme.Card.copy(alpha = 0.82f), RoundedCornerShape(14.dp))
            .padding(12.dp),
        verticalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Icon(
                Icons.Default.Memory, contentDescription = null,
                tint = Theme.Accent, modifier = Modifier.size(13.dp),
            )
            Text(
                " On-device · Melange",
                color = Theme.Accent, fontSize = 11.sp, fontWeight = FontWeight.Bold,
            )
            Spacer(Modifier.weight(1f))
            Text(
                if (mode == ClipFrameSource.Mode.BENCHMARK) "sustained" else "realtime",
                color = Theme.TextSecondary, fontSize = 10.sp, fontWeight = FontWeight.SemiBold,
            )
        }

        Row(modifier = Modifier.fillMaxWidth()) {
            Text(
                "memory",
                color = Theme.TextSecondary, fontSize = 11.sp, fontWeight = FontWeight.Medium,
            )
            Spacer(Modifier.weight(1f))
            Text(
                String.format(Locale.US, "%.0f MB · peak %.0f", stats.memoryMB, stats.peakMemoryMB),
                color = Theme.TextPrimary, fontSize = 11.sp, fontWeight = FontWeight.Medium,
            )
        }

        LatencyRow("YOLO26n", stats.detectMs)
        LatencyRow("RTMPose-s", stats.poseMs)
        if (liftAvailable) LatencyRow("3D lift", stats.liftMs)
        LatencyRow("pipeline", stats.totalMs, emphasized = true)
    }
}

@Composable
private fun LatencyRow(label: String, ms: Double, emphasized: Boolean = false) {
    Row(modifier = Modifier.fillMaxWidth()) {
        Text(
            label,
            color = if (emphasized) Theme.TextPrimary else Theme.TextSecondary,
            fontSize = 11.sp,
            fontWeight = if (emphasized) FontWeight.Bold else FontWeight.Medium,
        )
        Spacer(Modifier.weight(1f))
        Text(
            if (ms > 0) String.format(Locale.US, "%.1f ms", ms) else "—",
            color = if (emphasized) Theme.Accent else Theme.TextPrimary,
            fontSize = 11.sp,
            fontWeight = if (emphasized) FontWeight.Bold else FontWeight.Medium,
        )
    }
}
